### PR TITLE
`Text exercises`: Add preliminary AI feedback requests for students on text exercises using Athena

### PR DESCRIPTION
**Pull request by EneaGore**
> ONLY DEPLOY TO TS1 or TS2 or ma-schwind servers

#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
~- [ ] I added pre-authorization annotations according to the [guidelines]~(https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
~- [ ] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).~
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

Allow Preliminary Ai Generated Feedback for Students before the Exercise Due Date

### Description

A new option in Text Exercises that if enabled will allow students to request AI Feedback (with a pre defined limit of requests per student possible)
The results are saved as automatic feedback and attached to a submission.
The students can request feedback for each submission once.
Unlike before, where text exercises had one pereptually updated submission, now if there exists an athena feedback on a submission a new one will be created in case the student submits again.
Tutor assessment had to be slightly adjusted to ignore athena results, and also to retrieve the latest submission (instead of the default [0] which was the case until now)

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Text Exercise

1. Create New Exercise with the allow ai feedback feature enabled
2. (Optional) Ideally add some grading instructions
3. Participate as a student and request AI Feedback (limit must be 10 requests)
4. Confirm that each ai feedback request returns a result and creates a new submission
5. The student should be able to continue working on the submission
6. The results should always be visible, even before due date and before assessment due date
7. The results must be GRADED and tagged with PRELIMINARY on the detail view

8. Login as instructor.
9. Start an assessment.
10. Confirm that the latest submission is retrieved (You can check this in the participations or scores tab).
11. Confirm that assessment works as usually and that the student sees the final result.
12. Confirm that complaints and Complaint assessments work as usual.

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Screenshots

options
studentView
history
successAndOther
requireSubmitonResult

#### Screencast
video

## Summary by CodeRabbit

- **New Features**
	- Introduced automated feedback generation for text exercises via the new TextExerciseFeedbackService.
	- Enhanced filtering logic for displaying results based on assessment types, improving user experience.
	- Added new methods for retrieving student participation data with robust error handling.
	- Improved result display logic to indicate processing status for automatic feedback.
	- Refined logic for managing student participation results and feedback notifications in the Course Exercise Details component.
	- Updated submission handling to differentiate between new submissions and updates based on feedback.
	- Added localized messages for AI feedback status to enhance user communication.

- **Bug Fixes**
	- Enhanced handling of feedback notifications to ensure timely user alerts.

- **Tests**
	- Expanded test cases to validate the handling of results during automated assessments, ensuring accurate UI feedback.

### CODE REVIEW

File: src/main/webapp/app/exercises/shared/result/result.component.html
Comment (EneaGore) [reply src/main/webapp/app/exercises/shared/result/result.component.html:43]: I want specifically the case where its undefined which means that its still processing, if result.successful = false i do not want it to go in the condition.

File: src/main/java/de/tum/in/www1/artemis/service/TextExerciseFeedbackService.java
Comment (EneaGore) [reply src/main/java/de/tum/in/www1/artemis/service/TextExerciseFeedbackService.java:71]: There is an issue where most of the feedback correctly is recieved but at some point athena service throws an exception which makes the result.successful be false but request is proccessed and the feedback is visible to the client. Once we debug that i would reuse this line.

File: src/main/webapp/app/exercises/shared/result/result.component.html
Comment (EneaGore) [reply src/main/webapp/app/exercises/shared/result/result.component.html:43]: I specifically want only the case where its undefined. if its true or false i want it to not go into this condition.

File: src/main/webapp/app/exercises/shared/result/result.service.ts
Comment (EneaGore) [reply src/main/webapp/app/exercises/shared/result/result.service.ts:98]: Well i have not changed that part, but if irc the programming part also does some checks on programming specific things like build etc.

File: src/main/webapp/app/exercises/shared/feedback/feedback.component.html
Comment (EneaGore) [reply src/main/webapp/app/exercises/shared/feedback/feedback.component.html:93]: This condition did not allow to show the chart of points in the feedback detail modal view. Which assessment view do you mean exactly.

File: src/main/java/de/tum/in/www1/artemis/service/TextExerciseFeedbackService.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/TextExerciseFeedbackService.java:110]: Wouldn't it be enough to retrieve the latest submission from the database?
Please also use an `ElseThrow`-variant in the repository instead of checking here if a submission is present

File: src/main/java/de/tum/in/www1/artemis/service/TextExerciseFeedbackService.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/TextExerciseFeedbackService.java:156]:
```suggestion
            automaticResult.setScore(Math.clamp(totalFeedbacksScore, 0, 100));
```

File: src/main/java/de/tum/in/www1/artemis/service/TextExerciseFeedbackService.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/TextExerciseFeedbackService.java:159]:
```suggestion
            resultService.storeFeedbackInResult(automaticResult, feedbacks, true);
```

File: src/main/java/de/tum/in/www1/artemis/service/TextExerciseFeedbackService.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/TextExerciseFeedbackService.java:162]: This seams like a hacky solution. Submissions can have multiple results, e.g. programming exercises can have an automatic result as well as a manual assessment

File: src/main/java/de/tum/in/www1/artemis/service/TextSubmissionService.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/TextSubmissionService.java:101]:
```suggestion
        final var optionalParticipation = participationService.findOneByExerciseAndStudentLoginWithEagerSubmissionsAnyStateElseThrow(textExercise, user.getLogin());
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java:364]:
```suggestion
        Exercise exercise = exerciseRepository.findByIdElseThrow(exerciseId);
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java:373]: Please move this method or parts of it into the corresponding service

File: src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java:389]: Why can't you always use the studentParticipationRepository? The programmingExerciseParticipationService seams to also handle team exercises, why do you not handle them for text exercises as well?

File: src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java:393]: Why don't you load the needed attributes during the first database query?

File: src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java:396]: There are multiple differentiations between the different exercise types. I think it would be better to split the method into multiple parts, maybe even into different services (programmingExerciseService, ...)

File: src/main/java/de/tum/in/www1/artemis/web/rest/TextExerciseResource.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/web/rest/TextExerciseResource.java:431]:
```suggestion
```

File: src/main/webapp/app/exercises/shared/exercise-scores/exercise-scores.component.ts
Comment (JohannesStoehr) [inline src/main/webapp/app/exercises/shared/exercise-scores/exercise-scores.component.ts:176]:
```suggestion
            if (resultsWithoutAthena?.length != 0) {
```

File: src/main/webapp/app/exercises/shared/result/result.component.ts
Comment (JohannesStoehr) [inline src/main/webapp/app/exercises/shared/result/result.component.ts:309]: Please move this to the top

File: src/main/webapp/app/exercises/shared/result/result.service.ts
Comment (JohannesStoehr) [inline src/main/webapp/app/exercises/shared/result/result.service.ts:120]:
```suggestion
        aiFeedbackMessage = this.getResultStringNonProgrammingExercise(relativeScore, points, short);
        return `${aiFeedbackMessage} (${this.translateService.instant('artemisApp.result.preliminary')})`;
```
Or move all of it into the else branch, up to you...

File: src/main/webapp/app/exercises/shared/result/result.utils.ts
Comment (JohannesStoehr) [inline src/main/webapp/app/exercises/shared/result/result.utils.ts:294]: What about the ai feedback for programming exercises? The UI should be the same for both...

File: src/main/webapp/app/overview/exercise-details/course-exercise-details.component.ts
Comment (JohannesStoehr) [inline src/main/webapp/app/overview/exercise-details/course-exercise-details.component.ts:312]:
```suggestion
                        (changedParticipation.results?.length || 0) > (this.gradedStudentParticipation?.results?.length || 0)
```

File: src/main/webapp/app/overview/exercise-details/course-exercise-details.component.ts
Comment (JohannesStoehr) [inline src/main/webapp/app/overview/exercise-details/course-exercise-details.component.ts:317]:
```suggestion
                        (changedParticipation.results?.length || 0) > (this.gradedStudentParticipation?.results?.length || 0) &&
```

File: src/main/webapp/app/overview/exercise-details/course-exercise-details.component.ts
Comment (JohannesStoehr) [inline src/main/webapp/app/overview/exercise-details/course-exercise-details.component.ts:319]:
```suggestion
                        changedParticipation.results?.last()?.assessmentType === AssessmentType.AUTOMATIC_ATHENA &&
                        changedParticipation.results?.last()?.successful !== undefined
```

File: src/main/webapp/app/overview/exercise-details/course-exercise-details.component.ts
Comment (JohannesStoehr) [inline src/main/webapp/app/overview/exercise-details/course-exercise-details.component.ts:322]:
```suggestion
                        if (changedParticipation.results?.last()?.successful) {
```

File: src/main/webapp/app/overview/exercise-details/course-exercise-details.component.ts
Comment (JohannesStoehr) [inline src/main/webapp/app/overview/exercise-details/course-exercise-details.component.ts:324]:
```suggestion
                        } else if (!changedParticipation.results?.last()?.successful) {
```

File: src/main/webapp/app/overview/exercise-details/course-exercise-details.component.ts
Comment (JohannesStoehr) [inline src/main/webapp/app/overview/exercise-details/course-exercise-details.component.ts:311]:
```suggestion
```

File: src/main/webapp/app/overview/exercise-details/exercise-details-student-actions.component.html
Comment (JohannesStoehr) [inline src/main/webapp/app/overview/exercise-details/exercise-details-student-actions.component.html:258]: Why is it only disabled for exactly one submission? What happens if a student submits twice?

File: src/main/webapp/app/overview/exercise-details/exercise-details-student-actions.component.html
Comment (JohannesStoehr) [inline src/main/webapp/app/overview/exercise-details/exercise-details-student-actions.component.html:263]:
```suggestion
                            <span class="d-none d-md-inline" jhiTranslate="artemisApp.exerciseActions.requestAutomaticFeedback"></span>
```
You don't have to add this

File: src/main/java/de/tum/in/www1/artemis/service/TextExerciseFeedbackService.java
Comment (EneaGore) [reply src/main/java/de/tum/in/www1/artemis/service/TextExerciseFeedbackService.java:131]: yes

File: src/main/java/de/tum/in/www1/artemis/service/TextExerciseFeedbackService.java
Comment (EneaGore) [reply src/main/java/de/tum/in/www1/artemis/service/TextExerciseFeedbackService.java:90]: generateAutomaticNonGradedFeedback adds results to the participation, this is how it was done with programming so i kept it the same

File: src/main/webapp/app/overview/exercise-details/exercise-details-student-actions.component.html
Comment (EneaGore) [reply src/main/webapp/app/overview/exercise-details/exercise-details-student-actions.component.html:258]: I made it more general now for the last submission and also disabled if feedback is being generated

File: src/main/java/de/tum/in/www1/artemis/service/TextExerciseFeedbackService.java
Comment (EneaGore) [reply src/main/java/de/tum/in/www1/artemis/service/TextExerciseFeedbackService.java:147]: Well we wanted to have a new submission when AI Feedback is generated anyway, this doubles as an easy solution to not rewrite the TextAssessment.

File: src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java
Comment (EneaGore) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java:373]: Everything related to ParticipationResource will be improved on a follow up Pr which will also add athena feedback for modeling exercises.

**Review by FelixTJDietrich — CHANGES_REQUESTED** [review]
Here are a few UI improvements that would make it more consistent with the automatic testing for programming exercises:
- Don't show the timeline if it is just one result, see prog. ex.
- If you expand the results "Show all results", you are permanently showing a spinner even after the request is finished
- The color of the loading spinner and text is inconsistent with the prog. ex. where it is more blue

image

Some translation is missing, maybe we can also remove the `artemisApp.participation.errors.preconditions not met` alert and only show one:
image

File: src/main/java/de/tum/in/www1/artemis/service/TextExerciseFeedbackService.java
Comment (FelixTJDietrich) [inline src/main/java/de/tum/in/www1/artemis/service/TextExerciseFeedbackService.java:70]: Isn't this constant configurable in the course? @undernagruzez

File: src/main/java/de/tum/in/www1/artemis/service/TextExerciseFeedbackService.java
Comment (FelixTJDietrich) [inline src/main/java/de/tum/in/www1/artemis/service/TextExerciseFeedbackService.java:79]: programming exercise should be text exercise, check your comments again.

File: src/main/java/de/tum/in/www1/artemis/service/TextExerciseFeedbackService.java
Comment (FelixTJDietrich) [inline src/main/java/de/tum/in/www1/artemis/service/TextExerciseFeedbackService.java:93]: Here it is also programming exercise submission

File: src/main/java/de/tum/in/www1/artemis/service/TextExerciseFeedbackService.java
Comment (FelixTJDietrich) [inline src/main/java/de/tum/in/www1/artemis/service/TextExerciseFeedbackService.java:71]: Can this be localized on the client somehow? Also with the other errors
image

File: src/main/java/de/tum/in/www1/artemis/service/TextExerciseFeedbackService.java
Comment (EneaGore) [reply src/main/java/de/tum/in/www1/artemis/service/TextExerciseFeedbackService.java]: I reverted it to having a new submission be created only in the case that a student submit himself and the existing submission has a result. This means that submissions still work as they used to without athena feedback, that is the latest (only 1 in case of 0 results) submission gets edited.

**Comment by Jan-Thurner** [comment]
I will snatch the lock for TS1 if that's OK

**Review by coolchock — DISMISSED** [review]
tested on ts1 during testing session, works as expected

**Review by FelixTJDietrich — APPROVED** [review]
Code looks good to me, I think we can do the configurable feedback limit in a follow-up. Also some slight translation adjustments. And maybe hiding the button after the due date is over. 

Functionality-wise it works really good and translations work now. Also manual assessment still works as expected. Great job!

Screenshot 2024-09-04 at 15 31 45
Screenshot 2024-09-04 at 15 27 52
4-09-04 at 15 27 58](https://github.com/user-attachments/assets/1f2f682c-d1fa-4aed-bad5-e8649c41b581)
1-47b8-8b75-128290c0022b)

**Review by coolchock — APPROVED** [review]
re-approve after merging develop

**Review by FelixTJDietrich — DISMISSED** [review]

**Review by coolchock — DISMISSED** [review]
re-approve

**Comment by FelixTJDietrich** [comment]
Where is the additional server start coming from?
image

**Comment by EneaGore** [comment]
> Where is the additional server start coming from? image

It seems to be coming from the Intgration tets

**Review by FelixTJDietrich — APPROVED** [review]
Changes to tests LGTM and the additional server start is now gone

**Review by coolchock — APPROVED** [review]
re-approve after test changes

**Pull request merged**
